docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,433 @@
+   $collection->insertOne($person);
+
+   $person = $collection->findOne();
+
+The example above selects a collection and instructs it to use the ``PersonCodec`` for encoding and decoding documents.
+When inserting data, the ``PersonCodec`` is used to encode the document. When retrieving data, the ``PersonCodec`` is
+used to decode BSON data into a ``Person`` instance. Note that while the ``PersonCodec`` could technically decode any
+BSON document that contains a name field, we wouldn't want to use it for any other documents. Document codecs are meant
+to be used in a collection, or when decoding embedded documents.
+
+When working on a collection with a codec, the codec will only accept and return data of that type for certain
+operations. The ``bulkWrite``, ```findOneAndReplace``, ``insertMany``, ``insertOne``, and ``replaceOne`` operations
+will attempt to encode the given data using the provided codec. Trying to insert or replace a document that cannot be
+encoded will result in an exception. The ``aggregate``, ``find``, ``findOne``, ``findOneAndDelete``,
+``findOneAndReplace``, and ``findOneAndUpdate`` operations will attempt to decode returned documents using the provided
+codec. If the codec does not support the data returned, an exception will be thrown. You can disable codec usage for a
+specific operation or use a different codec (e.g. to decode the result of an aggregation pipeline) by specifying the
+nullable ``codec`` option for the operation. This will override the collection-level codec:
+
+.. code-block:: php
+
+   <?php
+
+   // Returns a document using the default type map
+   $collection->findOne(['name' => 'Jane Doe'], [
+       'codec' => null,
+   ]);
+
+   // Disables codec usage as the aggregate result will have a different format
+   $collection->aggregate([['$group' => [
+       '_id' => '$accountType',
+       'count' => ['$sum' => 1],
+   ]]], [
+       'codec' => null,
+   ]);
+
+Generic Codecs
+--------------
+
+The previous example showed how to define a codec for a specific class. However, sometimes you want to define codecs to
+handle a given type in all documents. For such codecs, you can implement the ``MongoDB\Codec\Codec`` interface. The
+following example defines a codec to store ``DateTimeInterface`` instances as BSON dates including the timezone:
+
+.. code-block:: php
+
+   <?php
+
+   final class DateTimeCodec implements MongoDB\Codec\Codec
+   {
+       // These traits define commonly used functionality to avoid duplication
+       use MongoDB\Codec\DecodeIfSupported;
+       use MongoDB\Codec\EncodeIfSupported;
+
+       public function canDecode($value): bool
+       {
+           // For maximum compatibility, this codec supports decoding both UTCDateTime instances and documents with a
+           // UTCDateTime instance and optional timezone
+           return
+               $value instanceof MongoDB\BSON\UTCDateTime ||
+               $value instanceof MongoDB\BSON\Document && $value->has('utc');
```

Comment:
> and documents with a UTCDateTime instance and optional timezone

I don't recall this coming up in any previous test suite examples. Isn't this likely to lead to problems if users want to use this codec alongside an existing DocumentCodec in CodecLibrary?

----

Maybe I'm misunderstanding this and the `MongoDB\BSON\Document` check here is actually necessary because you're going to encode a UTCDateTime into an embedded document. If so, I think the check here should consider _both_ fields in the embedded document: `utc` and `tz`.

And per my comment below, you might also do well to check their types here.

Reply:
The idea behind this is that it would handle any `UTCDateTime` value and any document with a `utc` property. Yes, that could clash with other codecs that use a similar structure, so we may want to add an extra field to be able to determine what this is.

On top of that, this presents an opportunity to once again mention that any `DocumentCodec` instances most likely shouldn't be in the default `CodecLibrary`, as they may pick up documents in error, and instead only using them directly in the `Codec` option. It also presents an opportunity to show how the functionality of an `odm` field used by the `Persistable` interface can be replicated.

Reply:
Updated to inspect the document as a whole and added a comment to explain that this can be slow for large documents.